R/settings.R

```diff
@@ -34,7 +34,13 @@ read_settings <- function(filename) {
   for (setting in names(default_settings)) {
     value <- get_setting(setting, config, default_settings)
     if (setting == "exclusions") {
-      value <- normalize_exclusions(value)
+      # value is a list ("filename" -> numeric, "filename")
+      # - `filename -> numeric` means exclude lines in `numeric` from `filename`
+      # - `filename` means exclude all lines in `filename`
+      # normalise_exclusions needs to know which directory the excluded-files
+      # are pinned against
+      dir_prefix <- if (is_directory(filename)) filename else NULL
```

Comment:
In my opinion, the behavior is still ambiguous if `filename` is not `NULL` and points to a file.

Could we use `dirname` in that case? Or would we want to assign `value <- NULL` in that case?

Let me know if that should not be thought of in this PR.

Reply:
I'm not really sure what the problem is. I've strived not to affect how linting works when `lint()` is called on single files.
But yeah, if lint() is called on a single file, then any exclusions defined in the config may not apply; so I guess syntastic, or similar, might flag lints when you're looking at a file that 'should be' excluded from linting. (but if files other than the file-of-interest are present in the exclusion list, there doesn't seem much point to NULL those out of the exlcusions list (given that you aren't linting them))

(Sorry if I missed your point)

Reply:
My concern is that it seems that `lint(filename)` looks at the configurations in a `.lintr` file in the same file pointed by `filename`.
So, if `filename` is an actual path (e.g. `inst/example/bad.R`), then `read_settings` might extract `exclusions: list("bad.R")` (assume we create a file `inst/example/.lintr` with that content).

In this situation, the results of
```R
lint("bad.R")
```
and
```R
lint("inst/example/bad.R")
```
will be different (assume the commands are run with the right workdir).

Looking at this part of the code in both executions we see that `dir_prefix = NULL` and `value = list("bad.R")`. Then, since the workdirs are different, the function `normalize_exclusions` will find the file `bad.R` in one case and not in the other, and the outputs will be different.

My suggestion, would be to make `dir_prefix` equal to the path to the folder of containing `filename`. E.g. one could change it to (or something nicer)
```R
dir_prefix <- if (is_directory(filename)) filename else {if (is.null(filename)) NULL else dirname(filename)}
```